libportal/notification.c

```diff
@@ -129,35 +272,30 @@ xdp_portal_add_notification (XdpPortal *portal,
                              GAsyncReadyCallback callback,
                              gpointer data)
 {
-  GAsyncReadyCallback call_done_cb = NULL;
-  CallDoneData *call_done_data = NULL;
+  GVariantBuilder *builder;
+  GVariantIter *iter;
+  g_autoptr(GTask) task = NULL;
 
   g_return_if_fail (XDP_IS_PORTAL (portal));
   g_return_if_fail (flags == XDP_NOTIFICATION_FLAG_NONE);
 
   ensure_action_invoked_connection (portal);
 
-  if (callback)
-    {
-      call_done_cb = call_done; 
-      call_done_data = g_new (CallDoneData, 1);
-      call_done_data->portal = g_object_ref (portal);
-      call_done_data->callback = callback,
-      call_done_data->data = data;
-    }
+  task = g_task_new (portal, cancellable, callback, data);
+  iter = g_variant_iter_new (notification);
+  builder = g_variant_builder_new (G_VARIANT_TYPE ("(sa{sv})"));
 
-  g_dbus_connection_call (portal->bus,
-                          PORTAL_BUS_NAME,
-                          PORTAL_OBJECT_PATH,
-                          "org.freedesktop.portal.Notification",
-                          "AddNotification",
-                          g_variant_new ("(s@a{sv})", id, notification),
-                          NULL,
-                          G_DBUS_CALL_FLAGS_NONE,
-                          -1,
-                          cancellable,
-                          call_done_cb,
-                          call_done_data);
+  g_variant_builder_add (builder, "s", id);
+  /* This will be closed once the async parsing is done */
+  g_variant_builder_open (builder, G_VARIANT_TYPE ("a{sv}"));
+
+  g_task_set_source_tag (task, xdp_portal_add_notification);
+  g_task_set_task_data (task, call_data_new (iter, builder), call_data_free);
```

Comment:
Doesn't change the semantics but iter and builder are both owned so maybe use g_autoptr/g_steal_pointer.